Resolve issues in NetApp E-Series Host module

lmprice opened the pull request:
##### SUMMARY
There are several defects that exist in the netapp_e_host module that this PR is intended to address.

There were several that have been reported with iSCSI, particularly the fact that iSCSI ports previously unknown to the array were not considered to be valid. In addition to removing this limitation (and allowing unknown ports to be mapped), several defects relating to the creation/updating of ports have been resolved with this patch.

There is still further work required with regard to the removal of ports, but that can be performed in a future change.

Unit tests and the skeleton for additional integration tests have been added to ensure that the module works and continues to work as designed.

This patch revamps the documentation and should greatly improve the user experience with the module. New examples have been provided and parameters more clearly defined.

I realize this patch may need some rework, but I wanted to get some other eyes on it before attempting to make any further changes.


<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
Bugfix Pull Request


##### COMPONENT NAME
netapp_e_host

##### ANSIBLE VERSION
```
ansible 2.6.0 (detached HEAD ce8c336d52) last updated 2018/04/18 11:23:03 (GMT -500)
  config file = None
  configured module search path = [u'/home/lorenp/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lorenp/ansible/lib/ansible
  executable location = /home/lorenp/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

It should no longer be necessary to include force_port to add ports that are not yet known to the storage-system. However, it is still necessary to do so when attempting to assign ports that already belong to a different host.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

mattclay commented on lib/ansible/modules/storage/netapp/Jenkinsfile:
```diff
@@ -0,0 +1,107 @@
+pipeline {
```
The Jenkinsfile should not be included.

lmprice replied:
I thought that I would try ;-) I wasn't sure what the policy was on stuff like that. I will remove it.

lmprice committed:
Resolve issues in NetApp E-Series Host module

The E-Series host module had some bugs relating to the update/creation
of host definitions when iSCSI initiators when included in the
configuration. This patch resolves this and other minor issues with
correctly detecting updates.

There were also several minor issues found that were causing issues with
truly idepotent updates/changes to the host definition.

This patch also provides some unit tests and integration tests to help
catch future issues in these areas.

fixes #28272

mattclay commented:
@lmprice If the `Jenkinsfile` can be made to work from another directory, we could consider placing it somewhere under the `hacking/` directory instead.

lmprice commented:
@mattclay Yes, the location of the Jenkinsfile within the project can certainly be customized. I think it would be useful at least for us to have present somewhere in the project.

lmprice commented:
@hulquest @broncofan Would you mind taking a look at this?

hulquest commented:
@lmprice I've been following along and just waiting for the green checks.  
shipit

broncofan commented on lib/ansible/modules/storage/netapp/netapp_e_host.py:
```diff
@@ -25,44 +24,141 @@
-    - name: Set Host Info
+    - name: Define or update an existing host named 'Host1'
+      netapp_e_host:
+        ssid: "1"
```
Did you mean to leave these hardcoded values in?

lmprice replied:
Yes. I thought that having real examples rather than variable references would be more useful. Although the IP address was real, everything else was intentionally faked.

broncofan commented:
shipit

lmprice commented:
@mattclay If there's a possibility of getting that Jenkinsfile added to the project somewhere in /hacking, I'm okay with doing that in a different patch.

Commit:
Improve NetApp E-Series Host module testing

The NetApp E-Series Host module integration test lacked feature test
verification to verify the changes made to the storage array.

The NetApp E-Series rest api was used to verify host create, update, and
remove changes made to the NetApp E-Series storage arrays.

lmprice commented:
@mattclay I'm a little behind on the overall process. What do I need to do to get this merged?

lmprice commented:
retest recheck

ndswartz reviewed (approved):
shipit

hulquest commented:
shipit

lmprice commented:
@mattclay Hey Matt, would you mind giving me a review on this one? I'd like to get some of the common code between these modules merged.
